### PR TITLE
Conformance profiles prototype

### LINKED ISSUE #1784: Initial conformance profiles and reporting prototype
This is a sub-issue of https://github.com/kubernetes-sigs/gateway-api/issues/1709 to start prototyping the GEP in it's current step so that we can test and try out some of the ideas we have so far and see how they work. The result that should be produced are non-released (not accessible by end-users of the go library unless _explicitly_ opted in to) tools for running profiles and producing a `ConformanceReport`.

### PR DESCRIPTION
**Pull request by mlavacca**
**What type of PR is this?**

/kind feature
/area conformance

**What this PR does / why we need it**:

This PR introduces the conformance profile machinery to verify compliance with given conformance profiles. It should be considered an implementation of https://github.com/kubernetes-sigs/gateway-api/blob/main/geps/gep-1709.md. The conformance profile test suite has been integrated and tested with Kong Kubernetes Ingress Controller, and below it is possible to see the outcome of such a test:

```json
{
  "implementation": {
    "organization": "Kong",
    "project": "Kubernetes Ingress Controller",
    "url": "https://github.com/Kong/kubernetes-ingress-controller",
    "version": "2.9.3",
    "contact": [
      "@mlavacca"
    ]
  },
  "date": "2023-04-24T13:24:05+02:00",
  "gatewayAPIVersion": "TODO",
  "profiles": [
    {
      "name": "HTTP",
      "core": {
        "result": "partial",
        "summary": "",
        "statistics": {
          "Passed": 26,
          "Skipped": 3,
          "Failed": 0
        },
        "skippedTests": [
          "GatewayModifyListeners",
          "HTTPRouteHeaderMatching",
          "GatewayWithAttachedRoutes"
        ]
      },
      "extended": {
        "status": {
          "result": "success",
          "summary": "",
          "statistics": {
            "Passed": 1,
            "Skipped": 0,
            "Failed": 0
          }
        },
        "supportedFeatures": [
          "HTTPRouteMethodMatching"
        ],
        "unsupportedFeatures": [
          "HTTPRouteSchemeRedirect",
          "HTTPRoutePathRedirect",
          "HTTPRouteHostRewrite",
          "HTTPRoutePathRewrite",
          "HTTPRouteQueryParamMatching",
          "HTTPResponseHeaderModification",
          "HTTPRoutePortRedirect"
        ]
      }
    },
    {
      "name": "TLS",
      "core": {
        "result": "partial",
        "summary": "",
        "statistics": {
          "Passed": 7,
          "Skipped": 3,
          "Failed": 0
        },
        "skippedTests": [
          "GatewayModifyListeners",
          "GatewayWithAttachedRoutes",
          "TLSRouteSimpleSameNamespace"
        ]
      }
    }
  ]
}
```

The result above has been produced by using the following settings:

```go
cSuite, err := suite.NewExperimentalConformanceTestSuite(suite.ExperimentalConformanceOptions{
  Options: suite.Options{
	  Client:               client,
	  GatewayClassName:     gatewayClass.Name,
	  Debug:                showDebug,
	  CleanupBaseResources: shouldCleanup,
	  BaseManifests:        conformanceTestsBaseManifests,
	  SupportedFeatures: sets.New(
		  // core
		  suite.SupportGateway,
		  suite.SupportReferenceGrant,
		  suite.SupportHTTPRoute,
		  suite.SupportTLSRoute,
  
		  // extended
		  suite.SupportHTTPRouteMethodMatching,
		  suite.SupportRouteDestinationPortMatching,
	  ),
	  SkipTests: []string{
		  // standard conformance
		  tests.HTTPRouteHeaderMatching.ShortName,
		  tests.GatewayWithAttachedRoutes.ShortName,
		  tests.GatewayModifyListeners.ShortName,
  
		  // TLSRoute
		  // https://github.com/Kong/kubernetes-ingress-controller/issues/3678
		  tests.TLSRouteSimpleSameNamespace.ShortName,
	  },
  },
  ConformanceProfiles: sets.New(
	  suite.HTTPConformanceProfileName,
	  suite.TLSConformanceProfileName,
  ),
  Implementation: confv1alpha1.Implementation{
	  Organization: "Kong",
	  Project:      "Kubernetes Ingress Controller",
	  URL:          "https://github.com/Kong/kubernetes-ingress-controller",
	  Version:      "2.9.3",
	  Contact:      []string{"@mlavacca"},
  },
})
assert.NoError(t, err)

cSuite.Setup(t)
cSuite.Run(t, tests.ConformanceTests)
report, err := cSuite.Report()
assert.NoError(t, err)
```

**Which issue(s) this PR fixes**:

Fixes #1784 

**Does this PR introduce a user-facing change?**:

```release-note
Addition of the conformance profiles. This feature enables an implementation to check compliance against a set of predetermined supported features and to produce a standard result (in the form of the `ConformanceProfile` API).
```

### CODE REVIEW

File: conformance/utils/suite/experimental_profiles.go
Comment (arkodg) [inline conformance/utils/suite/experimental_profiles.go:75]: can we reuse https://github.com/kubernetes-sigs/gateway-api/blob/387072cb29001a3771ff5e20ed5fde990749c6a6/conformance/utils/suite/features.go#L145 ?

File: conformance/utils/suite/experimental_reports.go
Comment (arkodg) [inline conformance/utils/suite/experimental_reports.go:123]: shouldnt the existence of unsupported tests be also treated as a `partial` result ?

**Comment by arkodg** [comment]
looks good @mlavacca, added some minor comments

File: conformance/utils/suite/experimental_reports.go
Comment (mlavacca) [reply conformance/utils/suite/experimental_reports.go:123]: We can see the extended features badge in two different ways:
1. Only implementations supporting ALL the extended features can have the `success` badge on them: implementations supporting only a subset of features with no failing test, get a `partial` badge
2. implementations supporting some extended features (even if not all of them) can have the `success` badge, along with the list of `supported/not-supported` features. In this scenario, no `partial` badge can be given, as either all the tests for a subset of tests pass, and the `success` badge is given, or some test fails, and the badge is `failure`.

In this implementation, we pursued the second way. I personally lean toward this implementation, as for extended tests we can select the specific features we are compliant with, so that an implementation can claim to be core conformant, and conformant with `x,y,z` extended features, and receive an according `success` badge.  I'm curious to hear your thoughts,  I think there is room for discussion here.

@shaneutt, any thoughts on this aspect?

File: conformance/utils/suite/experimental_profiles.go
Comment (mlavacca) [reply conformance/utils/suite/experimental_profiles.go]: Yes, it makes sense! I re-used that set here https://github.com/kubernetes-sigs/gateway-api/pull/1967/commits/7b7933191a43b5d6c404954958fa89997ed3b37a. I needed to slightly change the original sets as the original set therein declared was adding the HTTP core features as part of the extended features.

File: conformance/utils/suite/experimental_reports.go
Comment (arkodg) [reply conformance/utils/suite/experimental_reports.go:123]: so the issue here is reg readability, as a user I need to interpret `Result`  + `UnsupportedFeatures` to understand the true conformance level, Im proposing a tiered approach (`Result` encapsulates `UnsupportedFeatures`) so the user first interprets the `Result` and then looks at other fields to understand the WHY

Comment (mlavacca) [reply conformance/utils/suite/experimental_reports.go:123]: Here is the output for the extended section:
```json
"extended": {
        "status": {
          "result": "success",
          "summary": "",
          "statistics": {
            "Passed": 1,
            "Skipped": 0,
            "Failed": 0
          }
        },
        "supportedFeatures": [
          "HTTPRouteMethodMatching"
        ],
        "unsupportedFeatures": [
          "HTTPRouteSchemeRedirect",
          "HTTPRoutePathRedirect",
          "HTTPRouteHostRewrite",
          "HTTPRoutePathRewrite",
          "HTTPRouteQueryParamMatching",
          "HTTPResponseHeaderModification",
          "HTTPRoutePortRedirect"
        ]
      }
```
would you move supported and unsupported features somewhere else?

Comment (arkodg) [reply conformance/utils/suite/experimental_reports.go:123]: the current loc for `unsupportedFeatures` and `supportedFeatures` seems fine, Im having a hard time with `status.result` where `result` is the test output result which is `successful` is this case, but the `status` of the profile is `partial` (another synonym) since some features are unsupported

Comment (arkodg) [reply conformance/utils/suite/experimental_reports.go:123]: sure would be good to get another pair of eyes on this PR to make sure my concern is a minor one that can be addressed later based on user feedback

Comment (shaneutt) [reply conformance/utils/suite/experimental_reports.go:123]: /cc @skriss

File: conformance/apis/v1alpha1/conformanceprofile.go
Comment (mlavacca) [reply conformance/apis/v1alpha1/conformanceprofile.go:24]: yes, you are right. See a73d07cd070da453118903852a4cba26e86e2526

**Review by sunjayBhatia — COMMENTED** [review]
interesting data point (run with all supported features enabled and a few skipped tests):

```
{
  "implementation": {
    "organization": "projectcontour",
    "project": "contour",
    "url": "https://github.com/projectcontour/contour",
    "version": "foo",
    "contact": [
      "@bar"
    ]
  },
  "date": "2023-05-01T21:55:53Z",
  "gatewayAPIVersion": "TODO",
  "profiles": [
    {
      "name": "HTTP",
      "core": {
        "result": "failure",
        "summary": "",
        "statistics": {
          "Passed": 27,
          "Skipped": 1,
          "Failed": 1
        },
        "skippedTests": [
          "HTTPRouteRedirectHostAndStatus"
        ],
        "failedTests": [
          "GatewayObservedGenerationBump"
        ]
      },
      "extended": {
        "status": {
          "result": "partial",
          "summary": "",
          "statistics": {
            "Passed": 6,
            "Skipped": 2,
            "Failed": 0
          },
          "skippedTests": [
            "HTTPRouteRedirectScheme",
            "HTTPRouteRedirectPath"
          ]
        }
      }
    },
    {
      "name": "TLS",
      "core": {
        "result": "failure",
        "summary": "",
        "statistics": {
          "Passed": 9,
          "Skipped": 0,
          "Failed": 1
        },
        "failedTests": [
          "GatewayObservedGenerationBump"
...
```

looks like `GatewayObservedGenerationBump` (or any other common test between multiple profiles) shows up in two separate profiles, which I guess makes sense but could be a bit confusing when we go over the summary stats

**Review by keithmattix — APPROVED** [review]

**Review by Xunzhuo — APPROVED** [review]
LGTM, thanks for working on it @mlavacca !

**Comment by mlavacca** [comment]
> looks like GatewayObservedGenerationBump (or any other common test between multiple profiles) shows up in two separate profiles, which I guess makes sense but could be a bit confusing when we go over the summary stats

Yes, all the failing tests for a specific profile are enumerated in their own `failingTests` section. When multiple profiles share the same failing tests, we have the result you posted above. Where would you put such a piece of shared information to improve the interface readability?

File: conformance/utils/suite/experimental_reports.go
Comment (youngnick) [reply conformance/utils/suite/experimental_reports.go:123]: It's explicitly part of the API definition that Extended features are _optional_, if we required you to have zero unsupported extended features to be marked as passing conformance, then that would break that contract.

To put this another way, it is absolutely not intended that _every_ implementation meet every feature, and meeting all the extended features will just mean that you'll have a bunch of ticks in the eventual implementation comparison matrix. I strongly disagree that "success" should be gated on supporting everything.

**Comment by youngnick** [comment]
This LGTM, so I'll approve and hold, and wait for further maintainer review.

/approve
/hold

**Comment by shaneutt** [comment]
> This LGTM, so I'll approve and hold, and wait for further maintainer review.

It needs to be Rob:

/cc @robscott 

As I authored a lot of this PR so it would be very bold of me to review myself.

**Review by sunjayBhatia — APPROVED** [review]

File: conformance/utils/suite/experimental_profiles.go
Comment (robscott) [inline conformance/utils/suite/experimental_profiles.go:127]: I'm not sure why we need this workaround? We already have an exported `ConformanceTests` var, and each item in that list includes a list of `suite.SupportedFeature`. We also have `sets` that include the features in each profile (example: https://github.com/kubernetes-sigs/gateway-api/blob/main/conformance/utils/suite/features.go#L48). Relying on a prefix of the test name feels unnecessarily fragile and complex.

If I'm understanding this correctly, it means that each test can fit into exactly one bucket. Instead I think we want tests to be counted for every bucket they are relevant to.

File: conformance/utils/suite/experimental_profiles.go
Comment (robscott) [inline conformance/utils/suite/experimental_profiles.go:5]: Same comment applies in several places
```suggestion
Copyright 2023 The Kubernetes Authors.
```

File: conformance/utils/suite/experimental_suite.go
Comment (robscott) [inline conformance/utils/suite/experimental_suite.go:44]: As part of this PR, I think it would be helpful to have some documentation for how to actually run these tests. For example, it's not obvious to me how I'd fill in the implementation information below.

File: conformance/utils/suite/experimental_suite.go
Comment (robscott) [inline conformance/utils/suite/experimental_suite.go:203]: I may be missing something obvious, but what would lead to this being called multiple times? I'd assumed the primary way to run this would be via CLI. For reference, I'd assumed we'd have something similar to how Kubernetes conformance works where each implementation publishes the commands they ran to get the result they did: https://github.com/cncf/k8s-conformance/blob/master/v1.26/gke/README.md.

File: conformance/utils/suite/experimental_suite.go
Comment (robscott) [inline conformance/utils/suite/experimental_suite.go:170]: It would be helpful to understand how this is different from our main `Setup` function: https://github.com/kubernetes-sigs/gateway-api/blob/c67f3fd31976ea74ea06a78c2a88ae38df95a41d/conformance/utils/suite/suite.go#L147.

**Review by robscott — COMMENTED** [review]
Thanks for all the work on this @mlavacca!

File: conformance/utils/suite/experimental_suite.go
Comment (shaneutt) [reply conformance/utils/suite/experimental_suite.go:203]: The current suite simply does a `Run` and then it is done. This suite now has multiple knobs: you can run and then you can create a report from that run. Ultimately we don't _need_ to have this protection but for a low cost it makes the usage threadsafe in a future where there is more than one thing you would do with the object. I would suggest leaving this in unless there are some strong and specific objections because it makes the suite more robust.

File: conformance/utils/suite/experimental_suite.go
Comment (shaneutt) [reply conformance/utils/suite/experimental_suite.go:170]: It's just a copy that allows the experimental version. When this is no longer experimental we will merge the two suites.

Comment (robscott) [reply conformance/utils/suite/experimental_suite.go:170]: That makes sense, thanks! Adding some kind of comment in the code to describe this would be helpful for people that aren't as familiar with this plan.

File: conformance/utils/suite/experimental_suite.go
Comment (robscott) [reply conformance/utils/suite/experimental_suite.go:203]: That makes sense, I think I'd just come into this review expecting something primarily CLI-based and this comment comes with that bias. This does seem reasonable if these tests are primarily intended to be run with code. I still think it would be helpful to have as simple as possible a way for others to run tests against an implementation to verify the results that have been reported. Admittedly that's not really related to the lines of code this comment was on.

Comment (shaneutt) [reply conformance/utils/suite/experimental_suite.go:203]: I hear you on the CLI thing. I've created #1900 for this purpose, which is now a sub-issue of #1709.

File: conformance/utils/suite/experimental_suite.go
Comment (mlavacca) [reply conformance/utils/suite/experimental_suite.go:181]: I added some comments here 76dcc19

File: conformance/utils/suite/experimental_suite.go
Comment (mlavacca) [reply conformance/utils/suite/experimental_suite.go:47]: I added some indications in the code as comments. Were you thinking about something more structured, or is it enough? 76dcc19

File: conformance/utils/suite/experimental_profiles.go
Comment (mlavacca) [reply conformance/utils/suite/experimental_profiles.go]: I refactored the function, actually, that workaround wasn't really necessary. Please see 6393be5

**Comment by mlavacca** [comment]
@robscott Can you please take another look? All your comments should be addressed.

File: conformance/utils/suite/experimental_suite.go
Comment (robscott) [inline conformance/utils/suite/experimental_suite.go:91]: Tiny non-blocking nit: This feels more verbose than it needs to be and mentions concepts that we haven't actually defined yet (a profile badge). I think it would be fine if this just said `NewExperimentalConformanceTestSuite is a helper to use for creating a new ExperimentalConformanceTestSuite` or something similar.

**Review by robscott — COMMENTED** [review]
Thanks @mlavacca! Sorry for the delay on getting back to this one. This LGTM, have one tiny non-blocking nit but that can be handled in a follow up.

/lgtm
/approve
/hold cancel

File: conformance/utils/suite/experimental_suite.go
Comment (mlavacca) [reply conformance/utils/suite/experimental_suite.go]: I changed the comment as you suggested :)

**Comment by mlavacca** [comment]
@robscott there were conflicts, and the rebase removed LGTM and approved labels. Can you please approve it again?

**Pull request merged**
